File: packages/python/pezhwan/authorization.py
"""Authorization operations: RBAC roles, permissions, and ABAC can-checks."""

from __future__ import annotations

from typing import Any, Mapping, Optional, Sequence

from .client import PezhwanClient
from .errors import PezhwanApiError
from .models import Envelope


class Authorization:
    """Server-side RBAC/ABAC operations exposed by the identity server."""

    def __init__(self, client: PezhwanClient) -> None:
        self._client = client

    def assign_role(
        self,
        user_id: str,
        role_name: str,
        tenant_id: Optional[str] = None,
        application_id: Optional[str] = None,
    ) -> None:
        """Assign a role to a user within a tenant+application scope."""
        self._client.post(
            "/v1/admin/roles/assign",
            {
                "userId": user_id,
                "roleName": role_name,
                **({"tenantId": tenant_id} if tenant_id else {}),
                **({"applicationId": application_id} if application_id else {}),
            },
        )

    def remove_role(
        self,
        user_id: str,
        role_name: str,
        tenant_id: Optional[str] = None,
        application_id: Optional[str] = None,
    ) -> None:
        """Remove a role from a user within a tenant+application scope."""
        self._client.post(
            "/v1/admin/roles/remove",
            {
                "userId": user_id,
                "roleName": role_name,
                **({"tenantId": tenant_id} if tenant_id else {}),
                **({"applicationId": application_id} if application_id else {}),
            },
        )

    def get_user_roles(
        self,
        user_id: str,
        tenant_id: Optional[str] = None,
        application_id: Optional[str] = None,
    ) -> list[Mapping[str, Any]]:
        """Return the roles assigned to a user."""
        params = []
        if tenant_id:
            params.append(f"tenantId={tenant_id}")
        if application_id:
            params.append(f"applicationId={application_id}")
        suffix = f"?{'&'.join(params)}" if params else ""
        envelope = self._client.get(f"/v1/admin/users/{user_id}/roles{suffix}")
        return envelope.data or [] if hasattr(envelope, 'data') else []

    def get_user_permissions(
        self,
        user_id: str,
        tenant_id: Optional[str] = None,
        application_id: Optional[str] = None,
    ) -> list[str]:
        """Return the resolved permission names for a user."""
        params = []
        if tenant_id:
            params.append(f"tenantId={tenant_id}")
        if application_id:
            params.append(f"applicationId={application_id}")
        suffix = f"?{'&'.join(params)}" if params else ""
        envelope = self._client.get(f"/v1/admin/users/{user_id}/permissions{suffix}")
        data = envelope.data or [] if hasattr(envelope, 'data') else []
        return [str(p) for p in data]

    def can(
        self,
        user_id: str,
        permission: str,
        *,
        tenant_id: Optional[str] = None,
        application_id: Optional[str] = None,
        resource: Optional[str] = None,
        action: Optional[str] = None,
        resource_id: Optional[str] = None,
        attributes: Optional[Mapping[str, Any]] = None,
    ) -> bool:
        """Check if a user holds a given permission (RBAC + optional ABAC)."""
        body: dict[str, Any] = {
            "userId": user_id,
            "permission": permission,
        }
        if tenant_id:
            body["tenantId"] = tenant_id
        if application_id:
            body["applicationId"] = application_id
        if resource:
            body["resource"] = resource
        if action:
            body["action"] = action
        if resource_id:
            body["resourceId"] = resource_id
        if attributes:
            body["attributes"] = attributes
        envelope = self._client.post("/v1/admin/authorization/can", body)
        return bool((envelope.data or {}).get("allowed"))

File: packages/go/pezhwan/authorization.go
package pezhwan

import (
	"context"
	"net/url"
)

// AuthorizationService manages RBAC/ABAC operations via the identity server.
type AuthorizationService struct {
	client *Client
}

// AssignRole assigns a named role to a user within a tenant+application scope.
func (s *AuthorizationService) AssignRole(ctx context.Context, userID, roleName string, opts ...AuthorizationOption) error {
	body := map[string]any{"userId": userID, "roleName": roleName}
	for _, o := range opts {
		o(body)
	}
	_, err := s.client.post(ctx, "/v1/admin/roles/assign", body)
	return err
}

// RemoveRole removes a named role from a user within a tenant+application scope.
func (s *AuthorizationService) RemoveRole(ctx context.Context, userID, roleName string, opts ...AuthorizationOption) error {
	body := map[string]any{"userId": userID, "roleName": roleName}
	for _, o := range opts {
		o(body)
	}
	_, err := s.client.post(ctx, "/v1/admin/roles/remove", body)
	return err
}

// GetUserRoles returns the roles assigned to a user.
func (s *AuthorizationService) GetUserRoles(ctx context.Context, userID string, opts ...AuthorizationOption) ([]map[string]any, error) {
	data, err := s.client.get(ctx, "/v1/admin/users/"+userID+"/roles"+queryFor(opts))
	if err != nil {
		return nil, err
	}
	return stringifyList(data, "roles"), nil
}

// GetUserPermissions returns the resolved permission names for a user.
func (s *AuthorizationService) GetUserPermissions(ctx context.Context, userID string, opts ...AuthorizationOption) ([]string, error) {
	data, err := s.client.get(ctx, "/v1/admin/users/"+userID+"/permissions"+queryFor(opts))
	if err != nil {
		return nil, err
	}
	perms, _ := data["permissions"].([]any)
	out := make([]string, 0, len(perms))
	for _, p := range perms {
		if s, ok := p.(string); ok {
			out = append(out, s)
		}
	}
	return out, nil
}

// Can checks if a user holds a given permission (RBAC + optional ABAC).
func (s *AuthorizationService) Can(ctx context.Context, userID, permission string, opts ...AuthorizationOption) (bool, error) {
	body := map[string]any{"userId": userID, "permission": permission}
	for _, o := range opts {
		o(body)
	}
	data, err := s.client.post(ctx, "/v1/admin/authorization/can", body)
	if err != nil {
		return false, err
	}
	allowed, _ := data["allowed"].(bool)
	return allowed, nil
}

// AuthorizationOption mutates an authorization request body or query.
type AuthorizationOption func(map[string]any)

// WithAuthzTenantID sets the tenant scope.
func WithAuthzTenantID(tenantID string) AuthorizationOption {
	return func(b map[string]any) { b["tenantId"] = tenantID }
}

// WithAuthzApplicationID sets the application scope.
func WithAuthzApplicationID(appID string) AuthorizationOption {
	return func(b map[string]any) { b["applicationId"] = appID }
}

// WithAuthzResource sets the target resource and optional resource ID for ABAC checks.
func WithAuthzResource(resource string, resourceID string) AuthorizationOption {
	return func(b map[string]any) {
		b["resource"] = resource
		b["resourceId"] = resourceID
	}
}

// WithAuthzAction sets the target action for ABAC checks.
func WithAuthzAction(action string) AuthorizationOption {
	return func(b map[string]any) { b["action"] = action }
}

// WithAuthzAttributes sets the attribute map for ABAC checks.
func WithAuthzAttributes(attrs map[string]any) AuthorizationOption {
	return func(b map[string]any) { b["attributes"] = attrs }
}

// queryFor turns request-scoping options into a deterministic query string.
func queryFor(opts []AuthorizationOption) string {
	params := map[string]any{}
	for _, o := range opts {
		o(params)
	}
	values := url.Values{}
	for _, k := range []string{"tenantId", "applicationId"} {
		if v, ok := params[k].(string); ok && v != "" {
			values.Set(k, v)
		}
	}
	if len(values) == 0 {
		return ""
	}
	return "?" + values.Encode()
}

func stringifyList(data map[string]any, key string) []map[string]any {
	raw, _ := data[key].([]any)
	out := make([]map[string]any, 0, len(raw))
	for _, item := range raw {
		if m, ok := item.(map[string]any); ok {
			out = append(out, m)
		}
	}
	return out
}